c/certificateService.c

```diff
@@ -118,36 +119,83 @@ static void respondWithBadRequest(HttpResponse *response) {
 }
 
+
 static int serveMappingService(HttpService *service, HttpResponse *response)
 {
   HttpRequest *request = response->request;
 
-  if (!strcmp(request->method, methodPOST))
-  {
-    RUsermapParamList *userMapCertificateStructure 
-      = (RUsermapParamList*)safeMalloc31(sizeof(RUsermapParamList),"RUsermapParamList");
-    memset(userMapCertificateStructure, 0, sizeof(RUsermapParamList));
+  if (!strcmp(request->method, methodPOST)) {
+    RUsermapParamList userMapStructure = {0};
 
-    if(request->contentLength > sizeof(userMapCertificateStructure->certificate) || request->contentLength < 0) {
-      respondWithBadRequest(response);
+    int urlLength = strlen(request->uri);
+    if (urlLength < 0 || urlLength > 64) {
+      respondWithBadRequest(response, "URL length is not in range from 0 to 64 bytes.");
       return 0;
     }
+    char translatedURL[urlLength + 1];
+    memcpy(translatedURL, request->uri, strlen(request->uri));
+    a2e(translatedURL, sizeof(translatedURL));
+    char *found = strstr(translatedURL,"x509");
+
+    if (found != NULL) {
+    //  Certificate to user mapping
+      if (request->contentLength > sizeof(userMapStructure.certificate) || request->contentLength < 1) {
+        respondWithBadRequest(response, "The length of the certificate is longer than 4096 bytes");
+        return 0;
+      }
 
-    userMapCertificateStructure->certificateLength = request->contentLength;
-    memset(userMapCertificateStructure->certificate, 0, request->contentLength);
-    memcpy(userMapCertificateStructure->certificate, request->contentBody, request->contentLength);
+        userMapStructure.certificateLength = request->contentLength;
+        memset(userMapStructure.certificate, 0, request->contentLength);
+        memcpy(userMapStructure.certificate, request->contentBody, request->contentLength);
+
+        userMapStructure.functionCode = MAP_CERTIFICATE_TO_USERNAME;
+    } else {
+    //    Distinguished ID to user mapping
+        char *bodyNativeEncoding = copyStringToNative(request->slh, request->contentBody, request->contentLength);
+        char errorBuffer[2048];
+        Json *body = jsonParseUnterminatedString(request->slh, bodyNativeEncoding, request->contentLength, errorBuffer, sizeof(errorBuffer));
+        JsonObject *jsonObject = jsonAsObject(body);
+
+        char *distinguishedId = jsonObjectGetString(jsonObject, "dn");
+        if (distinguishedId == NULL || strlen(distinguishedId) == 0) {
+          respondWithBadRequest(response, "dn field not included in request body");
+          return 0;
+        } else if (strlen(distinguishedId) > sizeof(userMapStructure.distinguishedName)) {
+          respondWithBadRequest(response, "The length of the distinguished name is more than 246 bytes");
+          return 0;
+        }
+
+        userMapStructure.distinguishedNameLength = strlen(distinguishedId);
+        memset(userMapStructure.distinguishedName, 0, strlen(distinguishedId));
```

Comment:
This line seems to be useless because we overwrite the same amount of bytes on the next line. What did you intend here?